src/vmm/src/device_manager/mmio.rs

```diff
@@ -74,43 +78,72 @@ pub struct MMIODeviceInfo {
     pub irqs: Vec<u32>,
 }
 
+impl Aml for MMIODeviceInfo {
+    fn append_aml_bytes(&self, v: &mut Vec<u8>) {
+        let dev_id = self.irqs[0] - arch::IRQ_BASE;
+        debug!(
+            "acpi: Building AML for VirtIO device _SB_.V{:03}. memory range: {:#010x}:{} irq: {}",
+            dev_id, self.addr, self.len, self.irqs[0]
+        );
+        aml::Device::new(
+            format!("V{:03}", dev_id).as_str().into(),
+            vec![
+                &aml::Name::new("_HID".into(), &"LNRO0005"),
+                &aml::Name::new("_UID".into(), &dev_id),
+                &aml::Name::new("_CCA".into(), &aml::ONE),
+                &aml::Name::new(
+                    "_CRS".into(),
+                    &aml::ResourceTemplate::new(vec![
+                        &aml::Memory32Fixed::new(
+                            true,
+                            self.addr.try_into().unwrap(),
+                            self.len.try_into().unwrap(),
+                        ),
+                        &aml::Interrupt::new(true, true, false, false, self.irqs[0]),
```

Comment:
Mh, for these we assume that `self.irqs` is non-empty (and panic otherwise) by accessing index 0, but in line 320 (`register_mmio_boot_timer`), we describe the boot timer as an MMIO device with 0 irqs. How's that all fit together? I would expect the boottime will never show up in the ACPI tables, because its a fake device anyway (and also because all the tests pass). Does it make sense to handle the "no irqs" case here and emit a definition block without `_CRS`? Or should we add a check somewhere that prohibits MMIO devices with zero irqs (and then figure out what to do with the boot timer)?

Reply:
> Don't think we will hit a limit on 18 irq we have access to

What do you mean? We do not control what the users are doing. They might want to have 18 devices and we must have a very good reason to decrease that number (for example adding a device we really care about, such as VMGenID).

Reply:
I agree however that it is fine for `MMIODeviceInfo` to have up to 1 IRQ, something like `irq: Option<u32>`.

Reply:
`irq: Option<NonZeroU32>` will work even better, as we starts irq count from 5 (x86) and 32(aarch64).

Reply:
Removed the `Aml` trait implementation. Now using a dedicated function for VirtIO devices that receives as arguments the memory range and IRQ. Like that we can be sure at compile time that we did not accidentally created DSDT data for something other than a VirtIO device.